model_compression_toolkit/core/common/statistics_correction/compute_activation_bias_correction_of_graph.py

```diff
@@ -0,0 +1,210 @@
+# limitations under the License.
+# ==============================================================================
+from typing import List, Tuple, Any, Callable
+
+import numpy as np
+
+from model_compression_toolkit.core import CoreConfig
+from model_compression_toolkit.core.common import BaseNode, Graph
+from model_compression_toolkit.core.common.framework_implementation import FrameworkImplementation
+from model_compression_toolkit.core.common.framework_info import FrameworkInfo
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+
+
+def get_next_nodes_to_correct(node: BaseNode,
+                              graph: Graph,
+                              linear_node_types: NodeOperationMatcher,
+                              bypass_node_types: NodeOperationMatcher,
+                              bypass_nodes: List = None) -> Tuple[Any, Any]:
+    """
+    Search for the previous node which is not a bypass node of a given node. Go over the previous nodes of the node
+    and recursively search for a node.
+
+    Args:
+        node: Node to search for its previous node.
+        graph: Graph the node is in.
+        linear_node_types: Types of linear nodes to consider.
+        bypass_node_types: Types of nodes for bypassing to consider.
+        bypass_nodes: a list of bypass nodes found while running this function
+
+    Returns: The previous node (if found) and a list of bypass nodes (if any), or Nones if it were not found or there
+    are multiple incoming edges to one of nodes during the search (which means, the substitution can not be applied).
+    """
+
+    prev_nodes = graph.get_prev_nodes(node)
+
+    if len(prev_nodes) != 1:
+        return None, None  # pragma: no cover
+
+    prev_node = prev_nodes[0]
+
+    # If the previous node is not a bypass type, return it as the valid node along with any bypass nodes
+    if not bypass_node_types.apply(prev_node):
+        return prev_node, bypass_nodes
+
+    # If the previous node is a bypass node type, add it to the bypass_nodes list and continue searching
+    if bypass_node_types.apply(prev_node):
+        if bypass_nodes:
+            bypass_nodes.append(prev_node)
+        else:
+            bypass_nodes = [prev_node]
+        return get_next_nodes_to_correct(node=prev_node,
+                                         graph=graph,
+                                         linear_node_types=linear_node_types,
+                                         bypass_node_types=bypass_node_types,
+                                         bypass_nodes=bypass_nodes)
+    return None, None  # pragma: no cover
+
+
+def calculate_bin_centers(bin_edges: np.ndarray) -> np.ndarray:
+    """
+    Calculate the centers of bins given their edges.
+
+    Parameters:
+    bin_edges: Array of bin edges.
+
+    Returns:
+    np.ndarray: Array of bin centers.
+    """
+    # Ensure bin_edges is a numpy array
+    bin_edges = np.array(bin_edges, dtype=np.float32)
```

Comment:
if you're not sure of the argument type, state all type options in the type hints.
if you are sure it's numpy array, change to assert